ipc-queue/src/fifo.rs

```diff
@@ -0,0 +1,404 @@
+        let data = unsafe { &*self.data.as_ref().get() };
+        data.len()
+    }
+
+    fn offsets(&self) -> &AtomicUsize {
+        unsafe { self.offsets.as_ref() }
+    }
+
+    pub(crate) fn try_send_impl(&self, val: T) -> Result</*wake up reader:*/ bool, (TrySendError, T)> {
+        let (new, was_empty) = loop {
+            // 1. Load the current offsets.
+            let current = Offsets::new(self.offsets().load(Ordering::SeqCst), self.data_len() as u32);
+            let was_empty = current.is_empty();
+
+            // 2. If the queue is full, wait, then go to step 1.
+            if current.is_full() {
+                return Err((TrySendError::QueueFull, val));
+            }
+
+            // 3. Add 1 to the write offset and do an atomic compare-and-swap (CAS)
+            //    with the current offsets. If the CAS was not succesful, go to step 1.
+            let new = current.increment_write_offset();
+            let current = current.as_usize();
+            let prev = self.offsets().compare_and_swap(current, new.as_usize(), Ordering::SeqCst);
+            if prev == current {
+                break (new, was_empty);
+            }
+        };
+
+        // 4. Write the data, then the `id`.
+        let slot = self.slot(new.write_offset());
+        slot.copy_except_id(&val);
+        slot.set_id(val.get_id_non_atomic());
+
+        // 5. If the queue was empty in step 1, signal the reader to wake up.
+        Ok(was_empty)
+    }
+
+    pub(crate) fn try_recv_impl(&self) -> Result<(T, /*wake up writer:*/ bool), TryRecvError> {
+        // 1. Load the current offsets.
+        let current = Offsets::new(self.offsets().load(Ordering::SeqCst), self.data_len() as u32);
+        let was_full = current.is_full();
+
+        // 2. If the queue is empty, wait, then go to step 1.
+        if current.is_empty() {
+            return Err(TryRecvError::QueueEmpty);
+        }
+
+        // 3. Add 1 to the read offset.
+        let new = current.increment_read_offset();
+
+        let slot = loop {
+            // 4. Read the `id` at the new read offset.
+            let slot = self.slot(new.read_offset());
+            let id = slot.get_id();
+
+            // 5. If `id` is `0`, go to step 4 (spin). Spinning is OK because data is
+            //    expected to be written imminently.
+            if id != 0 {
+                break slot;
+            }
+        };
+
+        // 6. Read the data, then store `0` in the `id`.
+        let val = *slot;
+        slot.set_id(0);
+
+        // 7. Store the new read offset.
+        let after = fetch_adjust(
+            self.offsets(),
+            new.read as isize - current.read as isize,
+            Ordering::SeqCst,
+        );
+
+        // 8. If the queue was full in step 1, signal the writer to wake up.
+        //    ... or became full during read
+        let became_full = Offsets::new(after, self.data_len() as u32).is_full();
+        Ok((val, was_full || became_full))
```

Comment:
Ok I now think this makes sense. But I also think `became_full` is the only thing that should be checked. Please update the abi spec as well (“If the queue was full before step 7").

Reply:
Right, `was_full` implies `became_full`